File: OAuth/code_grant.py
import os
import sys
import base64
import json
import requests
import secrets
import threading
import webbrowser
import time

from utils.utils import ERROR_PAGE_SOURCE, app, run_server, HOST, PORT, SUCCESS_PAGE_SOURCE


def main():
    output_file = "config/ds_access_token.txt"
    api_account_id_file = "config/API_ACCOUNT_ID"

    # Check relative paths if files don't exist
    if not os.path.exists(output_file):
        output_file = "../config/ds_access_token.txt"
        api_account_id_file = "../config/API_ACCOUNT_ID"

    state = secrets.token_hex(5)
    client_id = os.getenv("INTEGRATION_KEY")
    client_secret = os.getenv("SECRET_KEY")
    target_account_id = os.getenv("TARGET_ACCOUNT_ID")

    authorization_endpoint = "https://account-d.docusign.com/oauth/"
    redirect_uri = f"http://{HOST}:{PORT}/authorization-code/callback"
    scope = "signature aow_manage"

    # Auth URL
    authorization_url = (
        f"{authorization_endpoint}auth?"
        f"response_type=code&scope={scope}"
        f"&client_id={client_id}"
        f"&state={state}"
        f"&redirect_uri={redirect_uri}"
    )

    print("\nOpen the following URL in a browser to continue:\n" + authorization_url + "\n")

    # Open browser
    webbrowser.open(authorization_url)

    # Start Flask server in a separate thread
    server_thread = threading.Thread(target=run_server, daemon=True)
    server_thread.start()

    # Wait for callback
    print("Waiting for authorization code...")

    from flask import render_template_string, request

    auth_data = {}

    @app.route("/authorization-code/callback", methods=["GET"])
    def handle_callback():
        nonlocal auth_data
        code = request.args.get("code")
        received_state = request.args.get("state")

        if code and received_state == state:
            auth_data = {"code": code, "state": received_state}

            return render_template_string(SUCCESS_PAGE_SOURCE, host=HOST, port=PORT)
        else:
            return render_template_string(ERROR_PAGE_SOURCE)

    # Wait until code is received
    while not auth_data:
        time.sleep(1)

    code = auth_data["code"]
    print("\nGetting an access token...\n")

    # Request access token
    token_url = authorization_endpoint + "token"
    auth_header = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()

    response = requests.post(
        token_url,
        data={
            "grant_type": "authorization_code",
            "redirect_uri": redirect_uri,
            "code": code,
        },
        headers={"Authorization": f"Basic {auth_header}"},
    )

    token_data = response.json()
    access_token = token_data.get("access_token")
    if not access_token:
        print("Error: failed to retrieve access token")
        sys.exit(1)

    # Save token
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    with open(output_file, "w") as f:
        f.write(access_token)
    print(f"\nAccess token has been written to {output_file}\n")

    # Get user info
    userinfo_url = authorization_endpoint + "userinfo"
    userinfo = requests.get(
        userinfo_url, headers={"Authorization": f"Bearer {access_token}"}
    ).json()

    if target_account_id and target_account_id != "{TARGET_ACCOUNT_ID}":
        target_account = next(
            (a for a in userinfo["accounts"] if a["account_id"] == target_account_id),
            None,
        )
        if not target_account:
            raise Exception(f"Targeted Account with Id {target_account_id} not found.")
        api_account_id = target_account["account_id"]
    else:
        default_account = next(
            (a for a in userinfo["accounts"] if a.get("is_default")), None
        )
        api_account_id = default_account["account_id"]

    # Save account ID
    with open(api_account_id_file, "w") as f:
        f.write(api_account_id)

    print(f"Account id: {api_account_id}")
    print(f"Account id has been written to {api_account_id_file}\n")


if __name__ == "__main__":
    main()

File: OAuth/code_grant.php
<?php

require './utils/utils.php';

$PORT                  = '8080';
$IP                    = 'localhost';

$outputFile            = 'config/ds_access_token.txt';
$state                 = bin2hex(random_bytes(5));

$clientID              = getenv("INTEGRATION_KEY");
$clientSecret          = getenv("SECRET_KEY");
$target_account_id     = getenv("TARGET_ACCOUNT_ID");

$authorizationEndpoint = 'https://account-d.docusign.com/oauth/';

$socket                =  'tcp://' . $IP . ':' . $PORT;
$redirectURI           = 'http://' . $IP . ':' . $PORT . '/authorization-code/callback';

$outputFile = "config/ds_access_token.txt";
$apiAccountIdFile = 'config/API_ACCOUNT_ID';
if (!file_exists($outputFile)) {
    $outputFile = "../config/ds_access_token.txt";
    $apiAccountIdFile = '../config/API_ACCOUNT_ID';
}

$scope = "signature aow_manage";

$authorizationURL = $authorizationEndpoint . 'auth?' . http_build_query(
    [
        'redirect_uri'  => $redirectURI,
        'scope'         => $scope,
        'client_id'     => $clientID,
        'state'         => $state,
        'response_type' => 'code'
    ]
);

echo "\nOpen the following URL in a browser to continue:\n" . $authorizationURL . "\n";


// Windows fix: https://stackoverflow.com/a/1327444/2226328
if (stripos(PHP_OS, 'WIN') === 0) {
    shell_exec('start "" "'.$authorizationURL.'"');
} else {
    shell_exec("xdg-open " . $authorizationURL);
}

$auth = startHttpServer($socket);

if ($auth['state'] != $state) {
    echo "\nWrong 'state' parameter returned\n";
    exit(2);
}

$code = $auth['code'];
echo "\nGetting an access token...\n";

$response = http(
    $authorizationEndpoint . 'token', [
        'grant_type'   => 'authorization_code',
        'redirect_uri' => $redirectURI,
        'code'         => $code
    ], [
        'Authorization: Basic ' . base64_encode($clientID . ':' .$clientSecret),
    ], true
);

$accessToken = $response->access_token;
file_put_contents($outputFile, $accessToken);
echo "\nAccess token has been written to " . $outputFile . "\n\n";

$userInfo = http($authorizationEndpoint . 'userinfo', false,
    [
        'Authorization: Bearer ' . $accessToken
    ]
);

if ($target_account_id != "{TARGET_ACCOUNT_ID}") {
    $targetAccountFound = false;
    foreach ($userInfo->accounts as $account_info) {
        if ($account_info->account_id == $target_account_id) {
            $APIAccountId = $account_info->account_id;
            $targetAccountFound = true;
            break;
        }
    }
    if (!$targetAccountFound) {
        throw new Exception("Targeted Account with Id " . $target_account_id . " not found.");
    }
} else {
    foreach ($userInfo->accounts as $account_info) {
        if ($account_info->is_default) {
            $APIAccountId = $account_info->account_id;
            break;
        }
    }
}

file_put_contents($apiAccountIdFile, $APIAccountId);
echo "Account id: $APIAccountId\n";
echo "Account id has been written to config/API_ACCOUNT_ID file...\n\n";
?>
